chardet/utf1632prober.py

```diff
@@ -0,0 +1,137 @@
+        if self.is_likely_utf16be():
+            return "utf-16be"
+        if self.is_likely_utf16le():
+            return "utf-16le"
+        # default to something valid
+        return "utf-16"
+
+    @property
+    def language(self):
+        return ""
+
+    def approx_32bit_chars(self):
+        return max(1.0, self.position / 4.0)
+
+    def approx_16bit_chars(self):
+        return max(1.0, self.position / 2.0)
+
+    def is_likely_utf32be(self):
+        approx_chars = self.approx_32bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            self.zeros_at_mod[0] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[1] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[2] / approx_chars > self.EXPECTED_RATIO and
+            self.nonzeros_at_mod[3] / approx_chars > self.EXPECTED_RATIO)
+
+    def is_likely_utf32le(self):
+        approx_chars = self.approx_32bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            self.nonzeros_at_mod[0] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[1] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[2] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[3] / approx_chars > self.EXPECTED_RATIO)
+
+    def is_likely_utf16be(self):
+        approx_chars = self.approx_16bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            (self.nonzeros_at_mod[1] + self.nonzeros_at_mod[3]) / approx_chars > self.EXPECTED_RATIO and
+            (self.zeros_at_mod[0] + self.zeros_at_mod[2]) / approx_chars > self.EXPECTED_RATIO)
+
+    def is_likely_utf16le(self):
+        approx_chars = self.approx_16bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            (self.nonzeros_at_mod[0] + self.nonzeros_at_mod[2]) / approx_chars > self.EXPECTED_RATIO and
+            (self.zeros_at_mod[1] + self.zeros_at_mod[3]) / approx_chars > self.EXPECTED_RATIO)
+
+    def feed(self, byte_str):
+        for c in byte_str:
+            if c == 0:
+                self.zeros_at_mod[self.position % 4] += 1
+            else:
+                self.nonzeros_at_mod[self.position % 4] += 1
+            self.position += 1
+        return self.state()
+
+    def state(self):
+        if self._state in [ProbingState.NOT_ME, ProbingState.FOUND_IT]:
+            # terminal, decided states
+            return self._state
+        elif self.get_confidence() > 0.95:
+            self._state = ProbingState.FOUND_IT
+        elif self.position > 4 * 1024:
+            # if we get to 4kb into the file, and we can't conclude it's UTF,
+            # let's give up
+            self._state = ProbingState.NOT_ME
+        return self._state
+
+    def get_confidence(self):
+        confidence = 0.99
+
+        if self.is_likely_utf16le() or self.is_likely_utf16be() or self.is_likely_utf32le() or self.is_likely_utf32be():
```

Comment:
I'd say if we look at even 100 bytes (25 code points) and they are all "0 0 0 x 0 0 0 y 0 0 0 z" etc - our confidence is likely to be sky high this is UTF32.  Likewise for UTF16.  

That was my thinking - I realise it's not particularly rigorous to state this without statistical analysis of many texts however, and maybe there are counter examples I'm not aware of.

One way to improve rigour, would be to verify exceptions to zeros are actually extended characters (surrogate pairs) of UTF16 also.  For instance in UTF16 you'd be looking for quads of (D8-DB) xx (DC-DF) xx to be recognised, e.g. - https://en.m.wikipedia.org/wiki/UTF-16#Examples

There are some other unused ranges in UTF16 and UTF32 also, so the detector could be made to flunk to zero if any of these are detected. 

You're making me think again about correctness of my implementation :)